api/src/main/java/com/datastrato/gravitino/exceptions/NotPartitionedTableException.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2024 Datastrato Pvt Ltd.
+ * This software is licensed under the Apache License version 2.
+ */
+package com.datastrato.gravitino.exceptions;
+
+import com.google.errorprone.annotations.FormatMethod;
+import com.google.errorprone.annotations.FormatString;
+
+/** Exception thrown when a partitioned table does not exist. */
+public class NotPartitionedTableException extends GravitinoRuntimeException {
```

Comment:
What's the exception used for? Can't exceptions `NoSuchTableException` or `UnsupportedOperationException` meet the scenario?

Reply:
`NoSuchTableException` means there is no such table at all. `UnsupportedOperationException` means we do not support some operations.
In this case, we already support partition management for Doris, if we operate some partition operations(like list, add, etc.) on non-partitioned table, `NotPartitionedTableException` should be thrown.

Reply:
I got you but I have a different view.

Partition operations are based on a table not a catalog:
```java
Table table = catalog.asTableCatalog().loadTable("tableName");
String[] partitionNames = table.supportPartitions().listPartitionNames();
```
And as the description of `supportPartitions()` API:
https://github.com/datastrato/gravitino/blob/445792f68df2d6f64a0b517103867719b9a1fc11/api/src/main/java/com/datastrato/gravitino/rel/Table.java#L74-L83

So the `UnsupportedOperationException` is for the non-partitioned table not for the catalog(Doris)

Reply:
> And as the description of supportPartitions() API:
[gravitino/api/src/main/java/com/datastrato/gravitino/rel/Table.java](https://github.com/datastrato/gravitino/blob/445792f68df2d6f64a0b517103867719b9a1fc11/api/src/main/java/com/datastrato/gravitino/rel/Table.java#L74-L83)

It is hard to determine whether a table is partitioned or not in `supportPartitions()` here, and the implementation of `supportPartitions()` just return an instance of `TableOperations` for both Hive and Doris catalog. According to this, is the definition of API here inappropriate? The `NotPartitionedTableException` is actually thrown in `listPartitionNames()`, `addPartition()` and so on, so this doesn't seem to conflict with the definition in the API either.

> So the UnsupportedOperationException is for the non-partitioned table not for the catalog(Doris)

 I'm ok with using `UnsupportedOperationException` for non-partitioned table, but the name of `NotPartitionedTableException` seems more specific to me. What do you think of it?

Reply:
> It is hard to determine whether a table is partitioned or not in supportPartitions() here

Can we use `ArrayUtils.isEmpty(partitioning())` to determine whether it's a partitioned table?

> I'm ok with using UnsupportedOperationException for non-partitioned table, but the name of 
NotPartitionedTableException seems more specific to me. What do you think of it?

Since the exception message can provide a specific reason, it seems unnecessary to introduce a new exception type. WDYT? @yuqi1129 @jerryshao

Reply:
> Can we use ArrayUtils.isEmpty(partitioning()) to determine whether it's a partitioned table?

That should work fine with Hive and Iceberg Table, but cannot be applied to JDBC tables. When we load tables from     JDBC catalogs, `partitioning` is not set actually, so `ArrayUtils.isEmpty(partitioning())` will always be true.

Reply:
> When we load tables from JDBC catalogs, partitioning is not set actually, so ArrayUtils.isEmpty(partitioning()) will always be true.

I am a little confused, if a JDBC table is a partitioned table, why isn't its partitioning set? Are you sure? If so, then there must be a bug somewhere.

Reply:
FYI: 
https://github.com/datastrato/gravitino/blob/d5d3fa25210e61f6bc425bc622bbdf5ee177db32/catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/JdbcCatalogOperations.java#L293-L301

https://github.com/datastrato/gravitino/blob/d5d3fa25210e61f6bc425bc622bbdf5ee177db32/catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java#L156-L194

I'm not sure why, maybe it's kind of complicated to get partition information when loading JDBC tables.

Reply:
I think it's a bug of JDBC-catalog. cc @yuqi1129 @FANNG1

Reply:
I do not know about it, @FANNG1 may have information.

Reply:
JDBCTableOperation is support Mysql and PG first which doesn't support partition(Gravitino MySQL catalog doesn't support).  we could extend the interface if support partition for doris.

Reply:
> JDBCTableOperation is support Mysql and PG first which doesn't support partition(Gravitino MySQL catalog doesn't support). we could extend the interface if support partition for doris.

Got that, maybe I could create an issue and fix it later. 

Back to this PR, should we introduce `NotPartitionedTableException` for non-partitioned table? What's your opinion? @yuqi1129

Reply:
I remember the following code being able to solve the problem. 
```
  default TableCatalog asTableCatalog() throws UnsupportedOperationException {
    throw new UnsupportedOperationException("Catalog does not support table operations");
  }
```

So, keeping `NotPartitionedTableException` is not so necessary.

Reply:
@mchades, `NotPartitionedTableException ` has been removed, and I create #4120 about the issue mentioned above, do you have any other suggestion about this PR?